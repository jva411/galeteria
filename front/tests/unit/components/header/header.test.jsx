/* eslint-disable no-undef */
import React from 'react';
import renderer from 'react-test-renderer';
import userEvent from '@testing-library/user-event';
import { render, screen } from '@testing-library/react';
import Header from '../../../../components/header/header';

describe('unit test of component \'header\'', () => {
    const handleLogout = jest.fn()
    afterEach(() => handleLogout.mockClear())

    const mockProps = {
        user: {
            role: 'ADMIN',
            username: 'Admin',
        },
        handleLogout: handleLogout
    }

    test('should render the username passed', () => {
        render(<Header {...mockProps} />)

        expect(screen.getByText(mockProps.user.username)).toBeInTheDocument()
    })

    test('should render the role passed', () => {
        render(<Header {...mockProps} />)

        expect(screen.getByText(mockProps.user.role)).toBeInTheDocument()
    })

    test('should render Avatar icon', () => {
        render(<Header {...mockProps} />)

        expect(screen.getByRole('img')).toBeInTheDocument()
    })

    test('should render notifications icon button when not pass hideNotifications prop', () => {
        render(<Header {...mockProps} />)

        expect(screen.getByRole('button', { name: /notifications/iu })).toBeInTheDocument()
    })

    test('should not render notifications icon button when pass hideNotifications prop', () => {
        render(<Header {...mockProps} hideNotifications />)

        expect(screen.queryByRole('button', { name: /notifications/iu })).not.toBeInTheDocument()
    })

    test('should render menu button', () => {
        render(<Header {...mockProps} />)

        expect(screen.getByRole('button', { name: /menu-button/iu })).toBeInTheDocument()
    })

    test('should not show menu list when the menu button wasn\'t clicked', () => {
        render(<Header {...mockProps} />)

        expect(screen.getByText(/logout/iu).parentElement.parentElement.style.visibility).toBe('hidden')
    })

    test('should show menu list when the menu button was clicked', () => {
        render(<Header {...mockProps} />)

        const button = screen.getByRole('button', { name: /menu.*button/iu })
        userEvent.click(button)

        expect(screen.getByText(/logout/iu).parentElement.parentElement.style.visibility).toBe('visible')
    })

    test('should call handle logout when logout button is clicked', () => {
        render(<Header {...mockProps} />)

        const logout = screen.getByText(/logout/iu)
        userEvent.click(logout)

        expect(logout).not.toBeVisible()
    })

    test('should Header match with Snapshot', () => {
        const tree = renderer.create(<Header {...mockProps} />).toJSON()

        expect(tree).toMatchSnapshot()
    })
})
