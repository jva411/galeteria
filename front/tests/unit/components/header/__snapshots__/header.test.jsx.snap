// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unit test of component 'header' should Header match with Snapshot 1`] = `
<div
  className="chakra-stack css-cxk74f"
>
  <div
    className="css-17xejub"
  />
  <div
    data-testid="notification-highlight"
    hidden={true}
  />
  <button
    aria-controls="popover-content-66"
    aria-expanded={false}
    aria-haspopup="dialog"
    className="chakra-button css-zse23x"
    id="popover-trigger-66"
    onClick={[Function]}
    title="notifications"
    type="button"
  >
    <svg
      aria-hidden={true}
      fill="currentColor"
      focusable={false}
      height="2.4rem"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 24 24"
      width="2.4rem"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M19 13.586V10c0-3.217-2.185-5.927-5.145-6.742C13.562 2.52 12.846 2 12 2s-1.562.52-1.855 1.258C7.185 4.074 5 6.783 5 10v3.586l-1.707 1.707C3.105 15.48 3 15.734 3 16v2c0 .553.447 1 1 1h16c.553 0 1-.447 1-1v-2c0-.266-.105-.52-.293-.707L19 13.586zM19 17H5v-.586l1.707-1.707C6.895 14.52 7 14.266 7 14v-4c0-2.757 2.243-5 5-5s5 2.243 5 5v4c0 .266.105.52.293.707L19 16.414V17zM12 22c1.311 0 2.407-.834 2.818-2H9.182C9.593 21.166 10.689 22 12 22z"
      />
    </svg>
  </button>
  <div
    className="chakra-popover__popper css-0"
    style={
      Object {
        "--popper-arrow-shadow-color": undefined,
        "--popper-arrow-size": "2rem",
        "minWidth": "max-content",
        "position": "absolute",
        "visibility": "hidden",
      }
    }
  >
    <section
      className="chakra-popover__content css-12f8fva"
      data-testid="popover-content"
      id="popover-content-66"
      onBlur={[Function]}
      onKeyDown={[Function]}
      role="dialog"
      style={
        Object {
          "opacity": 0,
          "transform": "scale(0.95) translateZ(0)",
          "transformOrigin": "var(--popper-transform-origin)",
          "visibility": "hidden",
        }
      }
      tabIndex={-1}
    />
  </div>
  <hr
    aria-orientation="vertical"
    className="chakra-divider css-2nwp48"
  />
  <div>
    <p>
      Admin
    </p>
    <p>
      ADMIN
    </p>
  </div>
  <button
    aria-controls="menu-list-68"
    aria-expanded={false}
    aria-haspopup="menu"
    className="chakra-button chakra-menu__menu-button css-zse23x"
    id="menu-button-68"
    onClick={[Function]}
    onKeyDown={[Function]}
    title="menu-button"
    type="button"
  >
    <svg
      aria-hidden={true}
      className="chakra-icon css-9xvt5q"
      focusable={false}
      viewBox="0 0 24 24"
    >
      <path
        d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
        fill="currentColor"
      />
    </svg>
  </button>
  <div
    className="css-0"
    style={
      Object {
        "minWidth": "max-content",
        "position": "absolute",
        "visibility": "hidden",
      }
    }
  >
    <div
      aria-orientation="vertical"
      className="chakra-menu__menu-list css-sjt5nk"
      id="menu-list-68"
      onKeyDown={[Function]}
      role="menu"
      style={
        Object {
          "opacity": 0,
          "transform": "scale(0.8) translateZ(0)",
          "transformOrigin": "var(--popper-transform-origin)",
          "visibility": "hidden",
        }
      }
      tabIndex={-1}
    >
      <button
        className="chakra-menu__menuitem css-171futd"
        id="menu-list-68-menuitem-69"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseMove={[Function]}
        role="menuitem"
        tabIndex={0}
        type="button"
      >
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="2"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M18.36 6.64a9 9 0 1 1-12.73 0"
          />
          <line
            x1="12"
            x2="12"
            y1="2"
            y2="12"
          />
        </svg>
        Logout
      </button>
    </div>
  </div>
  <span
    className="chakra-avatar css-b13ri1"
  >
    <div
      aria-label="Avatar"
      className="chakra-avatar__initials css-0"
      role="img"
    >
      A
    </div>
  </span>
</div>
`;
